var/spack/repos/builtin/packages/hdf5/package.py

```diff
@@ -70,23 +70,82 @@ class Hdf5(AutotoolsPackage):
     depends_on('szip', when='+szip')
     depends_on('zlib@1.1.2:')
 
-    @run_before('configure')
-    def validate(self):
-        """
-        Checks if incompatible variants have been activated at the same time
+    # According to ./configure --help thread-safe capabilities are:
+    # "Not compatible with the high-level library, Fortran, or C++ wrappers."
+    # (taken from hdf5@1.10.0patch1)
+    conflicts('+threadsafe', when='+cxx')
+    conflicts('+threadsafe', when='+fortran')
 
-        :param spec: spec of the package
-        :raises RuntimeError: in case of inconsistencies
+    @property
+    def libs(self):
+        """Hdf5 can be queried for the following parameters:
+        
+        - "hl": high-level interface
+        - "cxx": C++ APIs
+        - "fortran": fortran APIs
+        
+        :return: list of matching libraries
         """
+        query_parameters = self.spec.last_query.extra_parameters
+
+        shared = bool('+shared' in self.spec)
+
+        # This map contains a translation from query_parameters
+        # to the libraries needed
+        query2libraries = {
+            tuple(): ['libhdf5'],
+            ('cxx', 'fortran', 'hl'): [
```

Comment:
The order from client doesn't matter, because the key used to look-up this map is:
```python
key = tuple(sorted(query_parameters))
```

If the key is not in the map, a `KeyError` will be thrown. I don't know if we want to be more defensive than that: in C++ you "look before you leap", in Python you just "ask for forgiveness"...